src/objective-c/GRPCClient/private/GRPCChannelPool.m

```diff
@@ -0,0 +1,387 @@
+        [obj.callOptions.channelPoolDomain isEqualToString:_callOptions.channelPoolDomain]))
+    return NO;
+  if (!(obj.callOptions.channelID == _callOptions.channelID)) return NO;
+
+  return YES;
+}
+
+- (NSUInteger)hash {
+  NSUInteger result = 0;
+  result ^= _host.hash;
+  result ^= _callOptions.userAgentPrefix.hash;
+  result ^= _callOptions.responseSizeLimit;
+  result ^= _callOptions.compressAlgorithm;
+  result ^= _callOptions.enableRetry;
+  result ^= (unsigned int)(_callOptions.keepaliveInterval * 1000);
+  result ^= (unsigned int)(_callOptions.keepaliveTimeout * 1000);
+  result ^= (unsigned int)(_callOptions.connectMinTimeout * 1000);
+  result ^= (unsigned int)(_callOptions.connectInitialBackoff * 1000);
+  result ^= (unsigned int)(_callOptions.connectMaxBackoff * 1000);
+  result ^= _callOptions.additionalChannelArgs.hash;
+  result ^= _callOptions.PEMRootCertificates.hash;
+  result ^= _callOptions.PEMPrivateKey.hash;
+  result ^= _callOptions.PEMCertChain.hash;
+  result ^= _callOptions.hostNameOverride.hash;
+  result ^= _callOptions.transportType;
+  result ^= [_callOptions.logContext hash];
+  result ^= _callOptions.channelPoolDomain.hash;
+  result ^= _callOptions.channelID;
+
+  return result;
+}
+
+@end
+
+/**
+ * Time the channel destroy when the channel's calls are unreffed. If there's new call, reset the
+ * timer.
+ */
+@interface GRPCChannelCallRef : NSObject
+
+- (instancetype)initWithChannelConfiguration:(GRPCChannelConfiguration *)configuration
+                                destroyDelay:(NSTimeInterval)destroyDelay
+                               dispatchQueue:(dispatch_queue_t)dispatchQueue
+                              destroyChannel:(void (^)())destroyChannel;
+
+/** Add call ref count to the channel and maybe reset the timer. */
+- (void)refChannel;
+
+/** Reduce call ref count to the channel and maybe set the timer. */
+- (void)unrefChannel;
+
+@end
+
+@implementation GRPCChannelCallRef {
+  GRPCChannelConfiguration *_configuration;
+  NSTimeInterval _destroyDelay;
+  // We use dispatch queue for this purpose since timer invalidation must happen on the same
+  // thread which issued the timer.
+  dispatch_queue_t _dispatchQueue;
+  void (^_destroyChannel)();
+
+  NSUInteger _refCount;
+  NSTimer *_timer;
+}
+
+- (instancetype)initWithChannelConfiguration:(GRPCChannelConfiguration *)configuration
+                                destroyDelay:(NSTimeInterval)destroyDelay
+                               dispatchQueue:(dispatch_queue_t)dispatchQueue
+                              destroyChannel:(void (^)())destroyChannel {
+  if ((self = [super init])) {
+    _configuration = configuration;
+    _destroyDelay = destroyDelay;
+    _dispatchQueue = dispatchQueue;
+    _destroyChannel = destroyChannel;
+
+    _refCount = 0;
+    _timer = nil;
+  }
+  return self;
+}
+
+// This function is protected by channel pool dispatch queue.
+- (void)refChannel {
+  _refCount++;
```

Comment:
But it is supposed to be invoked in a serial queue, so I guess it is fine?

Reply:
It will be used that way, but its not clear from the interface that's required. For example, the passed in dispatchqueue is used in timerFire: to modify ivars, which is OK, but only because calls to ref/unref happen to happen on the same queue. It would be good to make that relationship explicit.

Assuming the passed in queue is serial (or arranging a serial queue that targets that queue) you could used dispatch_sync() to lock these accesses as well. Or you could just use regular locks like @synchronized.

Reply:
refactored.